File: androidApp/core/database/src/main/java/com/grateful/deadly/core/database/dao/BacklogDao.kt
package com.grateful.deadly.core.database.dao

import androidx.room.Dao
import androidx.room.Query
import androidx.room.Upsert
import com.grateful.deadly.core.database.entities.BacklogEntity
import kotlinx.coroutines.flow.Flow

/**
 * Backlog ("Up Next") data access. Ordering is by `position` ascending;
 * the head is the lowest-position live (non-tombstoned) row.
 *
 * Slice 1 keeps this purely local — no UI, no sync, no advance wiring. The
 * `deletedAt` tombstone is honored here (live queries filter it out) so the
 * later per-action sync (slice 4) can push removes without a schema change.
 */
@Dao
interface BacklogDao {

    /** Live backlog in play order (tombstones excluded). */
    @Query("SELECT * FROM backlog WHERE deletedAt IS NULL ORDER BY position ASC")
    fun observeBacklog(): Flow<List<BacklogEntity>>

    @Query("SELECT * FROM backlog WHERE deletedAt IS NULL ORDER BY position ASC")
    suspend fun getBacklog(): List<BacklogEntity>

    /** Head of the backlog (next to play), or null when empty. */
    @Query("SELECT * FROM backlog WHERE deletedAt IS NULL ORDER BY position ASC LIMIT 1")
    suspend fun peekHead(): BacklogEntity?

    @Query("SELECT * FROM backlog WHERE showId = :showId")
    suspend fun getById(showId: String): BacklogEntity?

    @Query("SELECT EXISTS(SELECT 1 FROM backlog WHERE showId = :showId AND deletedAt IS NULL)")
    suspend fun contains(showId: String): Boolean

    /** Largest position currently in use (live or tombstoned), or null when empty. */
    @Query("SELECT MAX(position) FROM backlog")
    suspend fun maxPosition(): Long?

    @Query("SELECT COUNT(*) FROM backlog WHERE deletedAt IS NULL")
    suspend fun count(): Int

    /** Insert or revive a row (re-adding a tombstoned show clears its tombstone). */
    @Upsert
    suspend fun upsert(entity: BacklogEntity)

    /** Tombstone a show (sync-friendly remove). */
    @Query("UPDATE backlog SET deletedAt = :deletedAt WHERE showId = :showId")
    suspend fun tombstone(showId: String, deletedAt: Long)

    @Query("UPDATE backlog SET position = :position WHERE showId = :showId")
    suspend fun setPosition(showId: String, position: Long)

    /** Tombstone every live row (Clear). */
    @Query("UPDATE backlog SET deletedAt = :deletedAt WHERE deletedAt IS NULL")
    suspend fun tombstoneAll(deletedAt: Long)

    /** Hard delete — used by tests/cleanup, not the user-facing remove. */
    @Query("DELETE FROM backlog WHERE showId = :showId")
    suspend fun hardDelete(showId: String)
}

File: iosApp/deadly/Core/Database/DAOs/BacklogDAO.swift
import GRDB

/// Backlog ("Up Next") data access. Ordering is by `position` ascending; the
/// head is the lowest-position live (non-tombstoned) row. Mirrors the Android
/// `BacklogRepository` op set. Slice 1: purely local — no UI, no sync, no advance.
struct BacklogDAO: Sendable {
    let database: AppDatabase

    // MARK: - Reads

    /// Live backlog in play order (head first), tombstones excluded.
    func fetchAll() throws -> [BacklogRecord] {
        try database.read { db in
            try BacklogRecord
                .filter(Column("deletedAt") == nil)
                .order(Column("position").asc)
                .fetchAll(db)
        }
    }

    /// Head of the backlog (next to play), or nil when empty.
    func peekHead() throws -> BacklogRecord? {
        try database.read { db in
            try BacklogRecord
                .filter(Column("deletedAt") == nil)
                .order(Column("position").asc)
                .fetchOne(db)
        }
    }

    /// A single row including tombstones — the push flusher reads this at push
    /// time to decide PUT (live) vs DELETE (tombstoned/absent).
    func fetchByIdIncludingTombstones(_ showId: String) throws -> BacklogRecord? {
        try database.read { db in
            try BacklogRecord.filter(Column("showId") == showId).fetchOne(db)
        }
    }

    /// Every local row including tombstones — used by sync reconcile to find
    /// local rows the server is missing or has an older copy of.
    func fetchAllIncludingTombstones() throws -> [BacklogRecord] {
        try database.read { db in
            try BacklogRecord.order(Column("position").asc).fetchAll(db)
        }
    }

    func contains(_ showId: String) throws -> Bool {
        try database.read { db in
            try BacklogRecord
                .filter(Column("showId") == showId && Column("deletedAt") == nil)
                .fetchCount(db) > 0
        }
    }

    func count() throws -> Int {
        try database.read { db in
            try BacklogRecord.filter(Column("deletedAt") == nil).fetchCount(db)
        }
    }

    // MARK: - Mutations

    /// Append to the bottom (or move an existing/tombstoned show to the bottom),
    /// clearing any tombstone.
    func addToBottom(_ showId: String, now: Int64) throws {
        try database.write { db in
            let maxPos = try Int64.fetchOne(db, sql: "SELECT MAX(position) FROM backlog") ?? 0
            try db.execute(
                sql: """
                    INSERT INTO backlog (showId, position, addedAt, updatedAt, deletedAt)
                    VALUES (?, ?, ?, ?, NULL)
                    ON CONFLICT(showId) DO UPDATE SET
                        position = excluded.position,
                        addedAt = excluded.addedAt,
                        updatedAt = excluded.updatedAt,
                        deletedAt = NULL
                    """,
                arguments: [showId, maxPos + 1, now, now]
            )
        }
    }

    /// Consume and return the head id (advance *commit*); nil when empty.
    func popHead(now: Int64) throws -> String? {
        try database.write { db in
            guard let head = try BacklogRecord
                .filter(Column("deletedAt") == nil)
                .order(Column("position").asc)
                .fetchOne(db) else { return nil }
            try db.execute(
                sql: "UPDATE backlog SET deletedAt = ?, updatedAt = ? WHERE showId = ?",
                arguments: [now, now, head.showId]
            )
            return head.showId
        }
    }

    /// Tombstone a single show (sync-friendly remove).
    func remove(_ showId: String, now: Int64) throws {
        try database.write { db in
            try db.execute(
                sql: "UPDATE backlog SET deletedAt = ?, updatedAt = ? WHERE showId = ?",
                arguments: [now, now, showId]
            )
        }
    }

    /// Rewrite the order to exactly `orderedShowIds` (drag-to-reorder).
    func reorder(_ orderedShowIds: [String], now: Int64) throws {
        try database.write { db in
            for (index, showId) in orderedShowIds.enumerated() {
                try db.execute(
                    sql: "UPDATE backlog SET position = ?, updatedAt = ? WHERE showId = ?",
                    arguments: [Int64(index), now, showId]
                )
            }
        }
    }

    /// Tombstone every live row (Clear).
    func clear(now: Int64) throws {
        try database.write { db in
            try db.execute(
                sql: "UPDATE backlog SET deletedAt = ?, updatedAt = ? WHERE deletedAt IS NULL",
                arguments: [now, now]
            )
        }
    }

    /// Write a row from a server pull verbatim (no push side effects). The LWW
    /// decision is made by the apply service before calling this.
    func applyFromSync(_ record: BacklogRecord) throws {
        try database.write { db in
            try db.execute(
                sql: """
                    INSERT INTO backlog (showId, position, addedAt, updatedAt, deletedAt)
                    VALUES (?, ?, ?, ?, ?)
                    ON CONFLICT(showId) DO UPDATE SET
                        position = excluded.position,
                        addedAt = excluded.addedAt,
                        updatedAt = excluded.updatedAt,
                        deletedAt = excluded.deletedAt
                    """,
                arguments: [record.showId, record.position, record.addedAt, record.updatedAt, record.deletedAt]
            )
        }
    }

    // MARK: - Observation

    func observeAll() -> ValueObservation<ValueReducers.Fetch<[BacklogRecord]>> {
        ValueObservation.tracking { db in
            try BacklogRecord
                .filter(Column("deletedAt") == nil)
                .order(Column("position").asc)
                .fetchAll(db)
        }
    }
}
